# **Predicting Arrests Following Terry Stops: Analyzing Factors Influencing Police Decision-Making**

## **Introduction**

In the realm of law enforcement, Terry Stops have been a subject of ongoing debate and scrutiny since their inception following the 1968 Supreme Court case Terry v. Ohio. These brief detentions, based on reasonable suspicion rather than probable cause, have become a common policing practice but have also raised concerns about potential biases and the fine line between public safety and individual rights. This project aims to delve into the complex dynamics of Terry Stops by leveraging data analytics and machine learning techniques. Our goal is to develop a predictive model that can determine the likelihood of an arrest following a Terry Stop, based on various factors such as the presence of weapons, time of day, and potentially sensitive demographic information like race and gender.  

By analyzing this data, we seek to uncover patterns and insights that could help law enforcement agencies refine their practices, address potential biases, and strike a balance between effective policing and fair treatment of all individuals. This analysis is particularly timely given the current national discourse on police reform and racial equity in the justice system. Our approach will involve careful consideration of ethical implications, transparent methodology, and a commitment to presenting findings in a way that contributes constructively to the ongoing dialogue about fair and effective law enforcement. Through this project, we aim to provide data-driven insights that can inform policy decisions, enhance police training, and ultimately contribute to building trust between law enforcement and the communities they serve. As we embark on this analysis, we recognize the sensitivity of the subject matter and the potential impact of our findings. Our objective is not to pass judgment but to illuminate patterns and trends that can lead to more informed, equitable, and effective policing practices in the context of Terry Stops.  

## **Business Understanding**

### Problem Statement

Law enforcement agencies are facing scrutiny over their stop-and-frisk practices, particularly regarding potential biases in arrest decisions. There's a need to understand the factors that influence whether an arrest is made following a Terry Stop to ensure fair and effective policing. This project aims to use data analysis and machine learning to shed light on the factors influencing arrest decisions during Terry Stops, potentially uncovering patterns that could help improve police practices and address concerns about bias in law enforcement.

### Objectives

- Develop a machine learning model to predict whether an arrest will be made after a Terry Stop.  
- Identify key factors that contribute to arrest decisions.  
- Analyze if and how demographic factors (race, gender) correlate with arrest outcomes.  
- Provide insights to help law enforcement agencies improve their decision-making processes and address potential biases.  

### Stakeholders

- Law enforcement agencies
- Police officers
- Policy makers
- Community leaders and civil rights organizations
- General public

### Data Sources

Dataset was obtained from Seattle Government which can be accessed using this [link](https://data.seattle.gov/Public-Safety/Terry-Stops/28ny-9ts8/data_preview)  
Terry Stops dataset containing information on:  

- Presence of weapons  
- Time of day of the call  
- Demographic information (race, gender) of subjects and officers  
- Arrest outcomes  

### Ethical Considerations

- Handling sensitive demographic data responsibly  
- Addressing potential biases in the dataset and model  
- Ensuring transparency in methodology and findings  
- Considering the broader societal implications of the analysis  

### Success Criteria 

- Develop a classifier with high accuracy in predicting arrests  
- Provide actionable insights for improving police practices  
- Contribute to the ongoing dialogue about fair policing and potential biases in law enforcement  

### Methodology

1. **Data Collection and Inspection**: Gathering the necessary data from the provided dataset.
2. **Data Cleaning and Preparation**: Cleaning the data to handle missing values, outliers, and incorrect data types.
3. **Exploratory Data Analysis (EDA)**: Analyzing the data to find patterns, relationships, and insights.
4. **Data Preprocessing**: This includes `Feature Selection`, `Target Variable Encoding`, `Time Conversion` and `Categorical Encoding`.
5. **Data Splitting**: Splitting data into Training and Testing sets using the **70/30** ratio.
6. **Model Selection and Training**: Come up with model choice and model training.
7. **Model Evaluation**: Check on Accuracy and Classification Report
8. **Discussion and Next Steps**: Model Performance Analaysis and Future work

## **Data Understanding**

Importing libraries

In [22]:
# import necessary libraries and functions
import pandas as pd
import numpy as np
from matplotlib.pyplot import plot
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler 
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score, roc_curve, roc_auc_score
from sklearn.preprocessing import OneHotEncoder
from sklearn import tree


Loading the Data and viewing the first 5 rows of the Dataset

In [23]:
# Load the dataset
df = pd.read_csv("./Data/Terry_Stops_20240823.csv")
# Viewing the first 5 rows
df.head()

Subject Age Group  Subject ID     GO / SC Num  Terry Stop ID  \
0           18 - 25          -1  20170000017766         230380   
1                 -          -1  20190000273690     8755198750   
2           36 - 45  7735709716  20200000186847    13469755649   
3           46 - 55          -1  20170000149189         460834   
4           46 - 55          -1  20160000001969         153868   

  Stop Resolution Weapon Type Officer ID  Officer YOB Officer Gender  \
0  Offense Report         NaN       6728         1973              M   
1   Field Contact           -       7528         1969              M   
2   Field Contact           -       6414         1964              M   
3          Arrest         NaN       5491         1967              M   
4   Field Contact         NaN       6899         1977              M   

                Officer Race  ...     Reported Time  \
0          Two or More Races  ...  08:03:00.0000000   
1                      White  ...  06:38:41.0000000   
2                      Asian  ...  01:27:52.0000000   
3  Black or African American  ...  09:53:00.0000000   
4                      White  ...  21:57:00.0000000   

                                   Initial Call Type  \
0                      ASLT - PERSON SHOT OR SHOT AT   
1  OBS WEAPN-IP/JO-GUN,DEADLY WPN (NO THRT/ASLT/D...   
2                 OBS - DOWN - CHECK FOR PERSON DOWN   
3             ASLT - WITH OR W/O WPNS (NO SHOOTINGS)   
4                                                  -   

                            Final Call Type Call Type  \
0         --DRIVE BY SHOOTING - NO INJURIES       911   
1       --WEAPON,PERSON WITH - OTHER WEAPON       911   
2  --SUSPICIOUS CIRCUM. - SUSPICIOUS PERSON       911   
3                         --ASSAULTS, OTHER       911   
4                                         -         -   

                                   Officer Squad Arrest Flag Frisk Flag  \
0                   SOUTH PCT 1ST W - R/S RELIEF           N          Y   
1                            NORTH PCT OPS - CPT           N          Y   
2      NORTH PCT 3RD W - NORA (JOHN) - PLATOON 1           N          N   
3  NORTH PCT 1ST W - LINCOLN (UNION) - PLATOON 1           N          Y   
4          NORTH PCT 3RD W - LINCOLN - PLATOON 1           N          N   

  Precinct Sector Beat  
0    South      S   S2  
1    North      U   U3  
2    North      N   N2  
3    North      L   L3  
4        -      -    -  

[5 rows x 23 columns]

Data Inspection

In [24]:
# The shape of the dataset
df.shape

(60984, 23)

The dataset contains 60,984 records and 23 columns.Here is a detailed list of the columns in the dataset, along with a brief description of each:

1. `Subject Age Group`: The age range of the subject involved in the Terry Stop (e.g., "18 - 25", "36 - 45").  
2. `Subject ID`: A unique identifier assigned to the subject of the Terry Stop.  
3. `GO / SC Num`: An internal case or record number associated with the incident.  
4. `Terry Stop ID`: A unique identifier for the specific Terry Stop event.  
5. `Stop Resolution`: The outcome or resolution of the Terry Stop (e.g., "Arrest", "Field Contact", "Offense Report").  
6. `Weapon Type`: The type of weapon found, if any, during the stop (e.g., "None", "Firearm", "Knife").  
7. `Officer ID`: A unique identifier assigned to the officer conducting the stop.  
8. `Officer YOB`: The year of birth of the officer involved in the stop.  
9. `Officer Gender`: The gender of the officer (e.g., "M" for Male, "F" for Female).  
10. `Officer Race`: The race or ethnicity of the officer (e.g., "White", "Black or African American", "Asian").  
11. `Subject Perceived Race`: The race or ethnicity of the subject as perceived by the officer (e.g., "White", "Black or African American", "Asian").  
12. `Subject Perceived Gender`: The gender of the subject as perceived by the officer (e.g., "Male", "Female").  
13. `Reported Date`: The date on which the Terry Stop was reported.  
14. `Reported Time`: The time at which the Terry Stop was reported (e.g., "08:03:00.0000000").  
15. `Initial Call Type`: The type of the initial call that led to the Terry Stop (e.g., "ASLT - PERSON SHOT OR SHOT AT").  
16. `Final Call Type`: The final classification of the call after the incident was resolved (e.g., "DRIVE BY SHOOTING - NO INJURIES").  
17. `Call Type`: The nature of the call, usually indicating whether it was a 911 call or another type (e.g., "911").  
18. `Officer Squad`: The squad or unit to which the officer belongs.  
19. `Arrest Flag`: Indicates whether an arrest was made during the stop (Y for Yes, N for No).  
20. `Frisk Flag`: Indicates whether a frisk or search was conducted during the stop (Y for Yes, N for No).  
21. `Precinct`: The police precinct where the Terry Stop occurred.  
22. `Sector`: A smaller geographical area within the precinct.  
23. `Beat`: The specific patrol area or beat within the sector.  

Key columns relevant for the binary classification task include:  

- `Stop Resolution`: Indicates the outcome of the stop (e.g., Arrest, Field Contact).  
- `Weapon Type`: Indicates whether a weapon was present.  
- `Officer Gender, Officer Race, Officer YOB`: Information about the officer.  
- `Subject Perceived Race, Subject Perceived Gender`: Information about the subject.  
- `Reported Time`: Time of the stop.  
- `Arrest Flag`: Indicates whether an arrest was made (Y or N).  

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60984 entries, 0 to 60983
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Subject Age Group         60984 non-null  object
 1   Subject ID                60984 non-null  int64 
 2   GO / SC Num               60984 non-null  int64 
 3   Terry Stop ID             60984 non-null  int64 
 4   Stop Resolution           60984 non-null  object
 5   Weapon Type               28419 non-null  object
 6   Officer ID                60984 non-null  object
 7   Officer YOB               60984 non-null  int64 
 8   Officer Gender            60984 non-null  object
 9   Officer Race              60984 non-null  object
 10  Subject Perceived Race    60984 non-null  object
 11  Subject Perceived Gender  60984 non-null  object
 12  Reported Date             60984 non-null  object
 13  Reported Time             60984 non-null  object
 14  Initial Call Type     

In [26]:
df.isna().sum()

Subject Age Group               0
Subject ID                      0
GO / SC Num                     0
Terry Stop ID                   0
Stop Resolution                 0
Weapon Type                 32565
Officer ID                      0
Officer YOB                     0
Officer Gender                  0
Officer Race                    0
Subject Perceived Race          0
Subject Perceived Gender        0
Reported Date                   0
Reported Time                   0
Initial Call Type               0
Final Call Type                 0
Call Type                       0
Officer Squad                 561
Arrest Flag                     0
Frisk Flag                      0
Precinct                        0
Sector                          0
Beat                            0
dtype: int64

From the Dataset information, `Weapon Type` has 32,565 null values and `Officer Squad` has 561 null values, the rest have no null values. In addition, majority of the column's DataTypes are `object`(**19**) whereas only **4** are `integers`  

In [27]:
# Checking for missing values and their percentages
missing_values = df.isnull().sum()
missing_percentage = (missing_values / df.shape[0]) * 100
missing_data = pd.DataFrame({'Missing Values': missing_values, 'Percentage': missing_percentage})
print(missing_data[missing_data['Missing Values'] > 0])

               Missing Values  Percentage
Weapon Type             32565   53.399252
Officer Squad             561    0.919913


We identified the columns with missing values and their respective percentages to understand the extent of missing data in our dataset. The results are as follows:  
| Column         | Missing Values |  Percentage    | 
|----------------|----------------|----------------|
| Weapon Type    |   32565        | 53.399252 %    |
| Officer Squad  |   561          | 0.919913 %     |

***Comment***  
Weapon Type: The high percentage of missing data for weapon type is a major concern. It could indicate issues in data collection, reporting, or recording processes. This missing data might introduce significant bias if not handled properly.  

Officer Squad: The low percentage of missing data for officer squad is less problematic. Since we are not going to use it anywhere. Nothing much is going to be done about it.

These missing values in Weapon Type column will need to be addressed during the data preprocessing stage to ensure the accuracy of our model.

In [28]:
#checking for duplicates in our dataset
df.duplicated().sum()

0

There are no dublicates in this Dataset.

In [29]:
# Summary statistics for numerical columns
df.describe()

Subject ID   GO / SC Num  Terry Stop ID   Officer YOB
count  6.098400e+04  6.098400e+04   6.098400e+04  60984.000000
mean   7.264174e+09  2.018662e+13   1.214806e+10   1984.078398
std    1.268858e+10  8.574292e+10   1.750952e+10      9.471936
min   -8.000000e+00 -1.000000e+00   2.802000e+04   1900.000000
25%   -1.000000e+00  2.017000e+13   2.389360e+05   1979.000000
50%   -1.000000e+00  2.018000e+13   5.088500e+05   1986.000000
75%    7.752509e+09  2.021000e+13   1.970754e+10   1991.000000
max    5.847921e+10  2.024000e+13   5.849088e+10   2002.000000

We got the summary statistics for numerical columns to understand the distribution and key characteristics of our dataset. below are some of the key observations:  

In this project, our key concern is the `Officer YOB`. From the summary, Majority of officers born between 1900 and 2002, their average YOB is 1984

In [30]:
# Checking for unique values for the Weapon Type and Officer Squad column which indicated there had missing values. 
# This will guide us on how to impute the values later.
weapon_type = df["Weapon Type"].unique()
officer_squad = df["Officer Squad"].unique()

print("Weapon Type Unique", weapon_type)
print("\n")
print("Officer Squad Unique", officer_squad)

Weapon Type Unique [nan '-' 'Lethal Cutting Instrument' 'Knife/Cutting/Stabbing Instrument'
 'Blunt Object/Striking Implement' 'Automatic Handgun' 'Other Firearm'
 'Firearm' 'Taser/Stun Gun' 'Mace/Pepper Spray' 'Firearm Other'
 'Club, Blackjack, Brass Knuckles' 'Handgun'
 'Personal Weapons (hands, feet, etc.)' 'Shotgun' 'Fire/Incendiary Device'
 'Rifle' 'None/Not Applicable' 'Club' 'Firearm (unk type)' 'Poison'
 'Brass Knuckles' 'Blackjack']


Officer Squad Unique ['SOUTH PCT 1ST W - R/S RELIEF' 'NORTH PCT OPS - CPT'
 'NORTH PCT 3RD W - NORA (JOHN) - PLATOON 1'
 'NORTH PCT 1ST W - LINCOLN (UNION) - PLATOON 1'
 'NORTH PCT 3RD W - LINCOLN - PLATOON 1' 'EAST PCT 3RD W - EDWARD'
 'WEST PCT 3RD W - MARY - PLATOON 1' 'WEST PCT 2ND W - SPECIAL BEATS'
 'EAST PCT 1ST W - E/G RELIEF (CHARLIE)'
 'NORTH PCT 2ND WATCH - NORTH BEATS' 'EAST PCT 3RD WATCH - CHARLIE RELIEF'
 'WEST PCT 2ND W - MARY BEATS' 'NORTH PCT 3RD W - L/U RELIEF'
 'WEST PCT 2ND W - DAVID - PLATOON 1' 'WEST PCT 3RD W - QUEEN - PLAT

## **Data Preprocessing**

### Handling Missing Values

We took the following steps to handle the missing values in the Weapon Type column:  

We first replaced all the "-" and NaN values with `Unknown` so as to indicate that Weapon might have been present at the scene only that the Type was not captured

In [31]:
# Step 1: Replace '-' with NaN
df['Weapon Type'].replace('-', pd.NA, inplace=True)

# Step 2: Impute NaN values (e.g., with 'Unknown')
df['Weapon Type'].fillna('Unknown', inplace=True)

In [37]:
# Viewing the first 5 rows to check on any missing values not identified as NaN
df.head()

Subject Age Group  Subject ID     GO / SC Num  Terry Stop ID  \
0           18 - 25          -1  20170000017766         230380   
1               NaN          -1  20190000273690     8755198750   
2           36 - 45  7735709716  20200000186847    13469755649   
3           46 - 55          -1  20170000149189         460834   
4           46 - 55          -1  20160000001969         153868   

  Stop Resolution Weapon Type Officer ID  Officer YOB Officer Gender  \
0  Offense Report     Unknown       6728         1973              M   
1   Field Contact     Unknown       7528         1969              M   
2   Field Contact     Unknown       6414         1964              M   
3          Arrest     Unknown       5491         1967              M   
4   Field Contact     Unknown       6899         1977              M   

                Officer Race  ...     Reported Time  \
0          Two or More Races  ...  08:03:00.0000000   
1                      White  ...  06:38:41.0000000   
2                      Asian  ...  01:27:52.0000000   
3  Black or African American  ...  09:53:00.0000000   
4                      White  ...  21:57:00.0000000   

                                   Initial Call Type  \
0                      ASLT - PERSON SHOT OR SHOT AT   
1  OBS WEAPN-IP/JO-GUN,DEADLY WPN (NO THRT/ASLT/D...   
2                 OBS - DOWN - CHECK FOR PERSON DOWN   
3             ASLT - WITH OR W/O WPNS (NO SHOOTINGS)   
4                                                NaN   

                            Final Call Type Call Type  \
0         --DRIVE BY SHOOTING - NO INJURIES       911   
1       --WEAPON,PERSON WITH - OTHER WEAPON       911   
2  --SUSPICIOUS CIRCUM. - SUSPICIOUS PERSON       911   
3                         --ASSAULTS, OTHER       911   
4                                       NaN       NaN   

                                   Officer Squad Arrest Flag Frisk Flag  \
0                   SOUTH PCT 1ST W - R/S RELIEF           N          Y   
1                            NORTH PCT OPS - CPT           N          Y   
2      NORTH PCT 3RD W - NORA (JOHN) - PLATOON 1           N          N   
3  NORTH PCT 1ST W - LINCOLN (UNION) - PLATOON 1           N          Y   
4          NORTH PCT 3RD W - LINCOLN - PLATOON 1           N          N   

  Precinct Sector Beat  
0    South      S   S2  
1    North      U   U3  
2    North      N   N2  
3    North      L   L3  
4      NaN    NaN  NaN  

[5 rows x 23 columns]

There are also other columns with "-" values. We are first going to convert them into NaN and re-check sum of null values in our dataset. There is a possibility they were not recorded when this data was being captured

In [33]:
# Replace '-' values with NaN
df.replace('-', np.nan, inplace=True)

# Check if there is any null values in the Dataset
df.isna().sum()

Subject Age Group            2200
Subject ID                      0
GO / SC Num                     0
Terry Stop ID                   0
Stop Resolution                 0
Weapon Type                     0
Officer ID                     24
Officer YOB                     0
Officer Gender                  0
Officer Race                    0
Subject Perceived Race       1816
Subject Perceived Gender      243
Reported Date                   0
Reported Time                   0
Initial Call Type           13473
Final Call Type             13473
Call Type                   13473
Officer Squad                 561
Arrest Flag                     0
Frisk Flag                    478
Precinct                    10619
Sector                      10770
Beat                        10764
dtype: int64

In [34]:
# Checking for missing values and their percentages
missing_values = df.isnull().sum()
missing_percentage = (missing_values / df.shape[0]) * 100
missing_data = pd.DataFrame({'Missing Values': missing_values, 'Percentage': missing_percentage})
print(missing_data[missing_data['Missing Values'] > 0])

                          Missing Values  Percentage
Subject Age Group                   2200    3.607504
Officer ID                            24    0.039355
Subject Perceived Race              1816    2.977830
Subject Perceived Gender             243    0.398465
Initial Call Type                  13473   22.092680
Final Call Type                    13473   22.092680
Call Type                          13473   22.092680
Officer Squad                        561    0.919913
Frisk Flag                           478    0.783812
Precinct                           10619   17.412764
Sector                             10770   17.660370
Beat                               10764   17.650531


Fom the above ouputs, it is evident the Dataset had lots of "-" values and after changing them to NaN, we can see that:  

- **Subject Age Group** has **2200** Null values which is at **3.6%**
- **Officer ID** has **24** Null values which is at **0.04%**
- **Subject Perceived Race** has **1816** Null values which is at **2.98%**
- **Subject Perceived Gender** has **243** Null values which is at **0.40%**
- **Initial Call Type**, **Final Call Type**, and **Call Type** have **13473** Null values each which are all at **22.09%** each
- **Officer Squad** has **561** Null values which is at **0.92%**
- **Frisk Flag** has **478** Null values which is at **0.78%**
- **Precinct** has **10619** Null values which is at **17.41%**
- **Sector** has **10770** Null values which is at **17.66%**
- **Beat** has **10764** Null values which is at **17.65%**


In [40]:
# For Initial Call Type, Final Call Type, Call Type: Impute with a placeholder
df["Initial Call Type"].fillna("Unknown", inplace=True)
df['Final Call Type'].fillna('Unknown', inplace=True)
df['Call Type'].fillna('Unknown', inplace=True)

# For Officer Squad: Impute with a placeholder
df['Officer Squad'].fillna('Unknown', inplace=True)

# For Precinct, Sector, Beat: Impute with 'Unknown' since these are categorical
df['Precinct'].fillna('Unknown', inplace=True)
df['Sector'].fillna('Unknown', inplace=True)
df['Beat'].fillna('Unknown', inplace=True)

# For Subject Perceived Race: Impute with mode
df['Subject Perceived Race'].fillna(df['Subject Perceived Race'].mode()[0], inplace=True)

# For Subject Perceived Gender: Impute with mode
df['Subject Perceived Gender'].fillna(df['Subject Perceived Gender'].mode()[0], inplace=True)

# For Officer ID: Impute with a placeholder since it's categorical
df['Officer ID'].fillna('Unknown', inplace=True)

# For Subject Age Group: Impute with mode (most common age group)
df['Subject Age Group'].fillna(df['Subject Age Group'].mode()[0], inplace=True)

# For Frisk Flag: Impute with mode (it’s binary Y/N)
df['Frisk Flag'].fillna(df['Frisk Flag'].mode()[0], inplace=True)

Imputed all the missing values and we are now going to confirm if their is any missing value

In [46]:
df.isna().any()

Subject Age Group           False
Subject ID                  False
GO / SC Num                 False
Terry Stop ID               False
Stop Resolution             False
Weapon Type                 False
Officer ID                  False
Officer YOB                 False
Officer Gender              False
Officer Race                False
Subject Perceived Race      False
Subject Perceived Gender    False
Reported Date               False
Reported Time               False
Initial Call Type           False
Final Call Type             False
Call Type                   False
Officer Squad               False
Arrest Flag                 False
Frisk Flag                  False
Precinct                    False
Sector                      False
Beat                        False
dtype: bool

We have completely worked out any missing values

In [35]:
# Columns to retain for X (features) and y (target)
relevant_columns = [
    'Stop Resolution', 'Weapon Type', 'Officer Gender', 
    'Officer Race', 'Officer YOB', 'Subject Perceived Race', 
    'Subject Perceived Gender', 'Reported Time', 'Frisk Flag', 
    'Arrest Flag'  # This is for y
]

In [36]:
# Keep only the relevant columns
terry_stops_df = df[relevant_columns]
terry_stops_df.isna().sum()


Stop Resolution                0
Weapon Type                    0
Officer Gender                 0
Officer Race                   0
Officer YOB                    0
Subject Perceived Race      1816
Subject Perceived Gender     243
Reported Time                  0
Frisk Flag                   478
Arrest Flag                    0
dtype: int64